README.md
# Axiod

Promise based HTTP client for Deno inspired by axios

## Features

- Make http requests from node.js
- Supports the Promise API
- Intercept request and response
- Transform request and response data
- Automatic transforms for JSON data
- Cancel requests [Waiting for deno support of Fetch Abort using signal and AbortController]

## Usage

```javascript
import axiod from "https://deno.land/x/axiod/mod.ts";

axiod.get("https://google.fr").then((response) => {
  // response
});
```

You can use type generics with Axiod

```typescript
import axiod from "https://deno.land/x/axiod/mod.ts";

const { data } = await axiod<{ delay: string }>(
  "https://postman-echo.com/delay/2"
);

// data type would be
// {delay: string}
```


Performing a `GET` request

```javascript
import axiod from "https://deno.land/x/axiod/mod.ts";

axiod
  .get("/user?ID=12345")
  .then((response) => {
    // handle success
    console.log(response);
  })
  .catch((error) => {
    // handle error
    console.log(error);
  })
  .then(() => {
    // always executed
  });

// Optionally the request above could also be done as
axiod
  .get("/user", {
    params: {
      ID: 12345,
    },
  })
  .then((response) => {
    console.log(response);
  })
  .catch((error) => {
    console.log(error);
  })
  .then(() => {
    // always executed
  });

// Want to use async/await? Add the `async` keyword to your outer function/method.
const getUser = () => {
  try {
    const response = await axiod.get("/user?ID=12345");
    console.log(response);
  } catch (error) {
    console.error(error);
  }
};
```

Performing a `POST` request

```javascript
import axiod from "https://deno.land/x/axiod/mod.ts";

axiod
  .post("/user", {
    firstName: "Fred",
    lastName: "Flintstone",
  })
  .then((response) => {
    console.log(response);
  })
  .catch((error) => {
    console.log(error);
  });
```

Performing multiple concurrent requests

```javascript
import axiod from "https://deno.land/x/axiod/mod.ts";

const getUserAccount = () => {
  return axiod.get("/user/12345");
};

const getUserPermissions = () => {
  return axiod.get("/user/12345/permissions");
};

Promise.all([getUserAccount(), getUserPermissions()]).then((results) => {
  const acct = results[0];
  const perm = results[1];
});
```

## Axiod API

Requests can be made by passing the relevant config to axiod.

**axiod(config)**
// Send a POST request

```javascript
import axiod from "https://deno.land/x/axiod/mod.ts";

axiod({
  method: "post",
  url: "/user/12345",
  data: {
    firstName: "Fred",
    lastName: "Flintstone",
  },
});
```

```javascript
import axiod from "https://deno.land/x/axiod/mod.ts";

// GET request for remote image in node.js
axiod({
  method: "get",
  url: "http://bit.ly/2mTM3nY",
  responseType: "stream",
}).then((response) => {
  response.data.pipe(fs.createWriteStream("ada_lovelace.jpg"));
});
```

**axiod(url[, config])**
// Send a GET request (default method)

```javascript
import axiod from "https://deno.land/x/axiod/mod.ts";

axiod("/user/12345");
```

### Request method aliases

For convenience aliases have been provided for all supported request methods.

```
axiod.request(config)
axiod.get(url[, config])
axiod.delete(url[, config])
axiod.head(url[, config])
axiod.options(url[, config])
axiod.post(url[, data[, config]])
axiod.put(url[, data[, config]])
axiod.patch(url[, data[, config]])
```

**NOTE**

> When using the alias methods url, method, and data properties don't need to be specified in config.

### Creating an instance

You can create a new instance of axiod with a custom config.

```javascript
import axiod from "https://deno.land/x/axiod/mod.ts";

// axiod.create([config]);
const instance = axiod.create({
  baseURL: "https://some-domain.com/api/",
  timeout: 1000,
  headers: { "X-Custom-Header": "foobar" },
});
```

## Request Config

These are the available config options for making requests. Only the url is required. Requests will default to GET if method is not specified.

```javascript
{
  // `url` is the server URL that will be used for the request
  url: '/user',

  // `method` is the request method to be used when making the request
  method: 'get', // default

  // `baseURL` will be prepended to `url` unless `url` is absolute.
  // It can be convenient to set `baseURL` for an instance of axiod to pass relative URLs
  // to methods of that instance.
  baseURL: 'https://some-domain.com/api/',

  // `headers` are custom headers to be sent
  headers: {'X-Requested-With': 'XMLHttpRequest'},

  // `params` are the URL parameters to be sent with the request
  // Must be a plain object or a URLSearchParams object
  params: {
    ID: 12345
  },

  // `paramsSerializer` is an optional function in charge of serializing `params`
  paramsSerializer: function (params) {
    return Qs.stringify(params, {arrayFormat: 'brackets'})
  },

  // `data` is the data to be sent as the request body
  // Only applicable for request methods 'PUT', 'POST', 'DELETE , and 'PATCH'
  // When no `transformRequest` is set, must be of one of the following types:
  // - string, plain object, ArrayBuffer, ArrayBufferView, URLSearchParams
  // - Browser only: FormData, File, Blob
  // - Node only: Stream, Buffer
  data: {
    firstName: 'Fred'
  },

  // syntax alternative to send data into the body
  // method post
  // only the value is sent, not the key
  data: 'Country=Brasil&City=Belo Horizonte',

  // `timeout` specifies the number of milliseconds before the request times out.
  // If the request takes longer than `timeout`, the request will be aborted.
  timeout: 1000, // default is `0` (no timeout)

  // `withCredentials` indicates whether or not cross-site Access-Control requests
  // should be made using credentials
  withCredentials: false, // default

  // `auth` indicates that HTTP Basic auth should be used, and supplies credentials.
  // This will set an `Authorization` header, overwriting any existing
  // `Authorization` custom headers you have set using `headers`.
  // Please note that only HTTP Basic auth is configurable through this parameter.
  // For Bearer tokens and such, use `Authorization` custom headers instead.
  auth: {
    username: 'janedoe',
    password: 's00pers3cret'
  }

  // `validateStatus` defines whether to resolve or reject the promise for a given
  // HTTP response status code. If `validateStatus` returns `true` (or is set to `null`
  // or `undefined`), the promise will be resolved; otherwise, the promise will be
  // rejected.
  validateStatus: function (status) {
    return status >= 200 && status < 300; // default
  },


  // `transformRequest` allows changes to the request data before it is sent to the server
  // This is only applicable for request methods 'PUT', 'POST', 'PATCH' and 'DELETE'
  // The last function in the array must return a string or an instance of Buffer, ArrayBuffer,
  // FormData or Stream
  // You may modify the headers object.
  transformRequest: [function (data, headers) {
    // Do whatever you want to transform the data

    return data;
  }],

  // `transformResponse` allows changes to the response data to be made before
  // it is passed to then/catch
  transformResponse: [function (data) {
    // Do whatever you want to transform the data

    return data;
  }],
}
```

## Response Schema

The response for a request contains the following information.

```javascript
{
  // `data` is the response that was provided by the server
  data: {},

  // `status` is the HTTP status code from the server response
  status: 200,

  // `statusText` is the HTTP status message from the server response
  statusText: 'OK',

  // `headers` the HTTP headers that the server responded with
  // All header names are lower cased and can be accessed using the bracket notation.
  // Example: `response.headers['content-type']`
  headers: {},

  // `config` is the config that was provided to `axiod` for the request
  config: {}
}
```

When using then, you will receive the response as follows:

```javascript
import axiod from "https://deno.land/x/axiod/mod.ts";

axiod.get("/user/12345").then((response) => {
  console.log(response.data);
  console.log(response.status);
  console.log(response.statusText);
  console.log(response.headers);
  console.log(response.config);
});
```

## Handling Errors

```javascript
import axiod from "https://deno.land/x/axiod/mod.ts";

axiod.get("/user/12345").catch((error) => {
  if (error.response) {
    // The request was made and the server responded with a status code
    // that falls out of the range of 2xx
    console.log(error.response.data);
    console.log(error.response.status);
    console.log(error.response.headers);
  } else {
    // Something happened in setting up the request that triggered an Error
    console.log("Error", error.message);
  }
  console.log(error.config);
});
```

## Testing

```javascript
// To test the module just run
deno test --import-map ./import_map.json --allow-net
```

## License

MIT


CHANGELOG.md
# Changelog

### 0.22.0 (October 01, 2021)

Fixes and Functionality:

- Added timeout support
- Added type generics support ( typescript )
- Fix falling test

Huge thanks to everyone who contributed to this release:

- [Maximilian5189](https://github.com/Maximilian5189)


api.test.ts
import { assertEquals } from "./mods/testing-assert.ts";
import axiod from "./mod.ts";

Deno.test("Axiod should have request method helpers", function () {
  assertEquals(typeof axiod.request, "function");
  assertEquals(typeof axiod.get, "function");
  assertEquals(typeof axiod.head, "function");
  assertEquals(typeof axiod.options, "function");
  assertEquals(typeof axiod.delete, "function");
  assertEquals(typeof axiod.post, "function");
  assertEquals(typeof axiod.put, "function");
  assertEquals(typeof axiod.patch, "function");
});

Deno.test("Axiod should have factory method", function () {
  assertEquals(typeof axiod.create, "function");
});

const instance = axiod.create();
Deno.test("Instance should have request methods", function () {
  assertEquals(typeof instance.request, "function");
  assertEquals(typeof instance.get, "function");
  assertEquals(typeof instance.options, "function");
  assertEquals(typeof instance.head, "function");
  assertEquals(typeof instance.delete, "function");
  assertEquals(typeof instance.post, "function");
  assertEquals(typeof instance.put, "function");
  assertEquals(typeof instance.patch, "function");
});

// Deno.test('should have interceptors', function () {
//   assertEquals(typeof instance.interceptors.request,'object');
//   assertEquals(typeof instance.interceptors.response,'object');
// });


timout.test.ts
import { assertEquals, assertThrowsAsync } from "./mods/testing-assert.ts";
import axiod from "./mod.ts";

Deno.test("Axiod with big timeout", async () => {
  const data = await axiod("https://postman-echo.com/delay/2", {
    timeout: 5000,
  });

  assertEquals(data.status, 200);
});

Deno.test("Axiod with tiny timeout", async () => {
  await assertThrowsAsync(async (): Promise<void> => {
    await axiod.get("https://postman-echo.com/delay/2", {
      timeout: 100,
    });
  });
});


generic.test.ts
import { assertEquals } from "./mods/testing-assert.ts";
import axiod from "./mod.ts";

Deno.test("Axiod using generic type", async () => {
  const { data } = await axiod<{ delay: string }>(
    "https://postman-echo.com/delay/2"
  );

  assertEquals(data.delay, "2");
});

Deno.test("Axiod.request using generic type", async () => {
  const { data } = await axiod.request<{ delay: string }>({
    url: "https://postman-echo.com/delay/2",
  });

  assertEquals(data.delay, "2");
});

Deno.test("Axiod.get using generic type", async () => {
  const { data } = await axiod.get<{ delay: string }>(
    "https://postman-echo.com/delay/2"
  );

  assertEquals(data.delay, "2");
});


Axiod.ts
import { urlJoin } from "url-join";

import type { IRequest, IConfig, Data, IAxiodResponse } from "./interfaces.ts";
import { methods } from "./helpers.ts";

function axiod(config: IRequest): Promise<IAxiodResponse> {
  return axiod.request(config);
}

axiod.defaults = {
  url: "/",
  method: "get",
  timeout: 0,
  withCredentials: false,
};

axiod.create = (config: IRequest) => {
  const instance = axiod.bind(config);
  instance.defaults = Object.assign({}, axiod.defaults, config);
  return instance;
};

axiod.request = ({
  url = "/",
  baseURL,
  method = "get",
  headers,
  params,
  data,
  timeout = 0,
  withCredentials = false,
  auth,
  paramsSerializer,
}: IRequest) => {
  // Url and Base url
  if (baseURL) {
    url = urlJoin(baseURL, url);
  }

  // Method
  if (methods.indexOf(method.toLowerCase().trim()) === -1) {
    throw new Error(`Method ${method} is not supported`);
  } else {
    method = method.toLowerCase().trim();
  }

  // Params
  let _params = "";
  if (params) {
    if (paramsSerializer) {
      _params = paramsSerializer(params);
    } else {
      _params = Object.keys(params)
        .map((key) => {
          return (
            encodeURIComponent(key) + "=" + encodeURIComponent(params[key])
          );
        })
        .join("&");
    }
  }

  // Add credentials to header
  if (withCredentials) {
    if (auth?.username && auth?.password) {
      if (!headers) {
        headers = {};
      }

      headers["Authorization"] =
        "Basic " +
        btoa(unescape(encodeURIComponent(`${auth.username}:${auth.password}`)));
    }
  }

  // Create fetch Request Config
  const fetchRequestObject: RequestInit = {};

  // Add method to Request Config
  if (method !== "get") {
    fetchRequestObject.method = method.toUpperCase();
  }

  // Add params to Request Config Url
  if (_params) {
    url = urlJoin(url, `?${params}`);
  }

  // Add body to Request Config
  if (data && method !== "get") {
    if (typeof data === "string" || data instanceof FormData) {
      fetchRequestObject.body = data;
    } else {
      try {
        fetchRequestObject.body = JSON.stringify(data);
        if (!headers) {
          headers = {};
        }

        headers["Accept"] = "application/json";
        headers["Content-Type"] = "application/json";
      } catch (ex) {}
    }
  }

  // Add headers to Request Config
  if (headers) {
    const _headers: Headers = new Headers();
    Object.keys(headers).forEach((header) => {
      if (headers && headers[header]) {
        _headers.set(header, headers[header]);
      }
    });
    fetchRequestObject.headers = _headers;
  }

  return fetch(url, fetchRequestObject).then(async (x) => {
    const _status: number = x.status;
    const _statusText: string = x.statusText;
    let _data: any = null;
    const contentType = x.headers.get("content-type") || "";
    if (contentType.toLowerCase().indexOf("json") === -1) {
      // Try to convert to json
      try {
        _data = await x.json();
      } catch (ex) {
        _data = await x.text();
      }
    } else {
      _data = await x.json();
    }
    const _headers: Headers = x.headers;
    const _config: IRequest = {
      url,
      baseURL,
      method,
      headers,
      params,
      data,
      timeout,
      withCredentials,
      auth,
      paramsSerializer,
    };

    if (_status >= 200 && _status < 300) {
      return Promise.resolve({
        status: _status,
        statusText: _statusText,
        data: _data,
        headers: _headers,
        config: _config,
      });
    } else {
      const error = {
        response: {
          status: _status,
          statusText: _statusText,
          data: _data,
          headers: _headers,
        },
        config: _config,
      };

      return Promise.reject(error);
    }
  });
};

axiod.get = (url: string, config?: IConfig) => {
  return axiod.request(Object.assign({}, { url }, config, { method: "get" }));
};
axiod.post = (url: string, data?: Data, config?: IConfig) => {
  return axiod.request(
    Object.assign({}, { url }, config, { method: "post", data })
  );
};
axiod.put = (url: string, data?: Data, config?: IConfig) => {
  return axiod.request(
    Object.assign({}, { url }, config, { method: "put", data })
  );
};
axiod.delete = (url: string, data?: Data, config?: IConfig) => {
  return axiod.request(
    Object.assign({}, { url }, config, { method: "delete", data })
  );
};
axiod.options = (url: string, data?: Data, config?: IConfig) => {
  return axiod.request(
    Object.assign({}, { url }, config, { method: "options", data })
  );
};
axiod.head = (url: string, data?: Data, config?: IConfig) => {
  return axiod.request(
    Object.assign({}, { url }, config, { method: "head", data })
  );
};
axiod.connect = (url: string, data?: Data, config?: IConfig) => {
  return axiod.request(
    Object.assign({}, { url }, config, { method: "connect", data })
  );
};
axiod.trace = (url: string, data?: Data, config?: IConfig) => {
  return axiod.request(
    Object.assign({}, { url }, config, { method: "trace", data })
  );
};
axiod.patch = (url: string, data?: Data, config?: IConfig) => {
  return axiod.request(
    Object.assign({}, { url }, config, { method: "patch", data })
  );
};

export default axiod;


helpers.ts
export const methods = [
  "get",
  "post",
  "put",
  "delete",
  "options",
  "head",
  "connect",
  "trace",
  "patch",
];

// /**
//    * Deep copy function for TypeScript.
//    * @param T Generic type of target/copied value.
//    * @param target Target value to be copied.
//    * @see Source project, ts-deeply https://github.com/ykdr2017/ts-deepcopy
//    * @see Code pen https://codepen.io/erikvullings/pen/ejyBYg
//    */
// export const deepCopy = <T>(target: T): T => {
//   if (target === null) {
//     return target;
//   }
//   if (target instanceof Date) {
//     return new Date(target.getTime()) as any;
//   }
//   // First part is for array and second part is for Realm.Collection
//   // if (target instanceof Array || typeof (target as any).type === 'string') {
//   if (typeof target === "object") {
//     if (typeof target[(Symbol as any).iterator] === "function") {
//       const cp = [] as any[];
//       if ((target as any as any[]).length > 0) {
//         for (const arrayMember of target as any as any[]) {
//           cp.push(deepCopy(arrayMember));
//         }
//       }
//       return cp as any as T;
//     } else {
//       const targetKeys = Object.keys(target);
//       const cp = {};
//       if (targetKeys.length > 0) {
//         for (const key of targetKeys) {
//           cp[key] = deepCopy(target[key]);
//         }
//       }
//       return cp as T;
//     }
//   }
//   // Means that object is atomic
//   return target;
// };


interfaces.ts
export interface IData {
  [key: string]: any;
}

export interface IHeaderData {
  [key: string]: any;
}

export type Data = string | IData | FormData;
export type TransformRequest = (data: Data, headers?: IHeaderData) => Data;
export type TransformResponse = (data: Data) => Data;

export interface IConfig {
  headers?: IHeaderData;
  params?: {
    [key: string]: string | number | boolean;
  };
  timeout?: number;
  withCredentials?: boolean;
  auth?: {
    username: string;
    password: string;
  };

  validateStatus?: Function;
  paramsSerializer?: Function;
  transformRequest?: Array<TransformRequest>;
  transformResponse?: Array<TransformResponse>;
}

export interface IRequest extends IConfig {
  url?: string;
  baseURL?: string;
  method?:
    | string
    | "get"
    | "post"
    | "put"
    | "delete"
    | "options"
    | "head"
    | "connect"
    | "trace"
    | "patch";
  data?: Data;
}

export interface IAxiodResponse<T = any> {
  status: number;
  statusText: string;
  data: T;
  headers: Headers;
  config: IRequest;
}


mod.ts
import { urlJoin } from "./mods/url-join.ts";

import type { Data, IAxiodResponse, IConfig, IRequest } from "./interfaces.ts";
import { methods } from "./helpers.ts";

function axiod<T = any>(
  url: string | IRequest,
  config?: IRequest
): Promise<IAxiodResponse<T>> {
  if (typeof url === "string") {
    return axiod.request(Object.assign({}, axiod.defaults, { url }, config));
  }
  return axiod.request(Object.assign({}, axiod.defaults, url));
}

axiod.defaults = {
  url: "/",
  method: "get",
  timeout: 0,
  withCredentials: false,
  validateStatus: (status: number) => {
    return status >= 200 && status < 300;
  },
};

axiod.create = (config?: IRequest) => {
  const instance = Object.assign({}, axiod);
  instance.defaults = Object.assign({}, axiod.defaults, config);
  instance.defaults.timeout = 1000;
  instance.request = (options: IRequest): Promise<IAxiodResponse> => {
    return axiod.request(Object.assign({}, instance.defaults, options));
  };
  instance.get = (url: string, config?: IConfig) => {
    return instance.request(
      Object.assign({}, { url }, config, { method: "get" })
    );
  };
  instance.post = (url: string, data?: Data, config?: IConfig) => {
    return instance.request(
      Object.assign({}, { url }, config, { method: "post", data })
    );
  };
  instance.put = (url: string, data?: Data, config?: IConfig) => {
    return instance.request(
      Object.assign({}, { url }, config, { method: "put", data })
    );
  };
  instance.delete = (url: string, data?: Data, config?: IConfig) => {
    return instance.request(
      Object.assign({}, { url }, config, { method: "delete", data })
    );
  };
  instance.options = (url: string, data?: Data, config?: IConfig) => {
    return instance.request(
      Object.assign({}, { url }, config, { method: "options", data })
    );
  };
  instance.head = (url: string, data?: Data, config?: IConfig) => {
    return instance.request(
      Object.assign({}, { url }, config, { method: "head", data })
    );
  };
  instance.connect = (url: string, data?: Data, config?: IConfig) => {
    return instance.request(
      Object.assign({}, { url }, config, { method: "connect", data })
    );
  };
  instance.trace = (url: string, data?: Data, config?: IConfig) => {
    return instance.request(
      Object.assign({}, { url }, config, { method: "trace", data })
    );
  };
  instance.patch = (url: string, data?: Data, config?: IConfig) => {
    return instance.request(
      Object.assign({}, { url }, config, { method: "patch", data })
    );
  };
  return instance;
};

axiod.request = <T = any>({
  url = "/",
  baseURL,
  method,
  headers,
  params,
  data,
  timeout,
  withCredentials,
  auth,
  validateStatus,
  paramsSerializer,
  transformRequest,
  transformResponse,
}: IRequest): Promise<IAxiodResponse<T>> => {
  // Url and Base url
  if (baseURL) {
    url = urlJoin(baseURL, url);
  }

  // Method
  if (method) {
    if (methods.indexOf(method.toLowerCase().trim()) === -1) {
      throw new Error(`Method ${method} is not supported`);
    } else {
      method = method.toLowerCase().trim();
    }
  } else {
    method = "get";
  }

  // Params
  let _params = "";
  if (params) {
    if (paramsSerializer) {
      _params = paramsSerializer(params);
    } else {
      _params = Object.keys(params)
        .map((key) => {
          return (
            encodeURIComponent(key) + "=" + encodeURIComponent(params[key])
          );
        })
        .join("&");
    }
  }

  // Add credentials to header
  if (withCredentials) {
    if (auth?.username && auth?.password) {
      if (!headers) {
        headers = {};
      }

      headers["Authorization"] =
        "Basic " +
        btoa(unescape(encodeURIComponent(`${auth.username}:${auth.password}`)));
    }
  }

  // Create fetch Request Config
  const fetchRequestObject: RequestInit = {};

  // Add method to Request Config
  if (method !== "get") {
    fetchRequestObject.method = method.toUpperCase();
  }

  // Add params to Request Config Url
  if (_params) {
    url = urlJoin(url, `?${_params}`);
  }

  // Add body to Request Config
  if (data && method !== "get") {
    // transformRequest
    if (
      transformRequest &&
      Array.isArray(transformRequest) &&
      transformRequest.length > 0
    ) {
      for (var i = 0; i < (transformRequest || []).length; i++) {
        if (transformRequest && transformRequest[i]) {
          data = transformRequest[i](data, headers);
        }
      }
    }

    if (
      typeof data === "string" ||
      data instanceof FormData ||
      data instanceof URLSearchParams
    ) {
      fetchRequestObject.body = data;
    } else {
      try {
        fetchRequestObject.body = JSON.stringify(data);
        if (!headers) {
          headers = {};
        }

        headers["Accept"] = "application/json";
        headers["Content-Type"] = "application/json";
      } catch (ex) {}
    }
  }

  // Add headers to Request Config
  if (headers) {
    const _headers: Headers = new Headers();
    Object.keys(headers).forEach((header) => {
      if (headers && headers[header]) {
        _headers.set(header, headers[header]);
      }
    });
    fetchRequestObject.headers = _headers;
  }

  // Timeout
  const controller = new AbortController();
  fetchRequestObject.signal = controller.signal;

  let timeoutCounter: number = 0;

  if ((timeout || 0) > 0) {
    timeoutCounter = setTimeout(() => {
      timeoutCounter = 0;
      controller.abort();
    }, timeout);
  }

  // Start request
  return fetch(url, fetchRequestObject).then(async (x) => {
    // Clear timeout
    if (timeoutCounter) {
      clearTimeout(timeoutCounter);
    }

    const _status: number = x.status;
    const _statusText: string = x.statusText;

    // Data
    let _data: any = null;

    // Check content type and then do the needed transformations
    const contentType = x.headers.get("content-type") || "";
    if (contentType.toLowerCase().indexOf("json") === -1) {
      // Try to convert to json
      try {
        _data = await x.clone().json();
      } catch (ex) {
        _data = await x.clone().text();
      }
    } else {
      _data = await x.clone().json();
    }

    // transformResponse
    if (transformResponse) {
      if (
        transformResponse &&
        Array.isArray(transformResponse) &&
        transformResponse.length > 0
      ) {
        for (var i = 0; i < (transformResponse || []).length; i++) {
          if (transformResponse && transformResponse[i]) {
            _data = transformResponse[i](_data);
          }
        }
      }
    }

    const _headers: Headers = x.headers;
    const _config: IRequest = {
      url,
      baseURL,
      method,
      headers,
      params,
      data,
      timeout,
      withCredentials,
      auth,
      paramsSerializer,
    };

    // Validate the status code
    let isValidStatus = true;

    if (validateStatus) {
      isValidStatus = validateStatus(_status);
    } else {
      isValidStatus = _status >= 200 && _status < 300;
    }

    if (isValidStatus) {
      return Promise.resolve({
        status: _status,
        statusText: _statusText,
        data: _data,
        headers: _headers,
        config: _config,
      });
    } else {
      const error = {
        response: {
          status: _status,
          statusText: _statusText,
          data: _data,
          headers: _headers,
        },
        config: _config,
      };

      return Promise.reject(error);
    }
  });
};

axiod.get = <T = any>(url: string, config?: IConfig) => {
  return axiod.request<T>(
    Object.assign({}, { url }, config, { method: "get" })
  );
};
axiod.post = <T = any>(url: string, data?: Data, config?: IConfig) => {
  return axiod.request<T>(
    Object.assign({}, { url }, config, { method: "post", data })
  );
};
axiod.put = <T = any>(url: string, data?: Data, config?: IConfig) => {
  return axiod.request<T>(
    Object.assign({}, { url }, config, { method: "put", data })
  );
};
axiod.delete = <T = any>(url: string, data?: Data, config?: IConfig) => {
  return axiod.request<T>(
    Object.assign({}, { url }, config, { method: "delete", data })
  );
};
axiod.options = <T = any>(url: string, data?: Data, config?: IConfig) => {
  return axiod.request<T>(
    Object.assign({}, { url }, config, { method: "options", data })
  );
};
axiod.head = <T = any>(url: string, data?: Data, config?: IConfig) => {
  return axiod.request<T>(
    Object.assign({}, { url }, config, { method: "head", data })
  );
};
axiod.connect = <T = any>(url: string, data?: Data, config?: IConfig) => {
  return axiod.request<T>(
    Object.assign({}, { url }, config, { method: "connect", data })
  );
};
axiod.trace = <T = any>(url: string, data?: Data, config?: IConfig) => {
  return axiod.request<T>(
    Object.assign({}, { url }, config, { method: "trace", data })
  );
};
axiod.patch = <T = any>(url: string, data?: Data, config?: IConfig) => {
  return axiod.request<T>(
    Object.assign({}, { url }, config, { method: "patch", data })
  );
};

export default axiod;
